### ISSUE TITLE
[binding-http] Sending the first GET request to observe a property starts logging of the TD to the console

### ISSUE THREAD

**Issue by egekorkan** [eclipse/thingweb.node-wot#244]
## Version

core: 0.7.0
binding-http: 0.7.0
installed via npm

## Behavior

On the Thing side, every second I write to a property that updates its value with a function like this:

```js
    private pollSensors(){
        setInterval(()=>{
            bme280.readSensorData().then((data) => {
                
                this.thing.writeProperty("temperature",data.temperature_C)
                this.thing.writeProperty("pressure",data.pressure_hPa)

            });
        },1000)
    }
```
When the Thing starts, it logs the sensor values every second like the following

```
ExposedThing 'RainbowHAT' directly sets Property 'temperature' to value '36.81'
ExposedThing 'RainbowHAT' directly sets Property 'pressure' to value '949.9067119739695'
```
So far so good. Sending an HTTP GET request to read a property (`/properties/temperature`) results in the TD being logged to the console. This is weird but maybe it is OK. 

However, sending a GET request to observe a property (`/properties/temperature/observable`) results in the TD being logged every second or every time the property is written to.

**Comment by danielpeintner** [comment]
@egekorkan May I ask you to provide the log snippet. Does it show "who" logs it (e.g., http-client.ts or consumed-hting.ts ). I do have difficulties finding the "source" of the problem...

This is another good reason why we should also implement the proposal from https://github.com/eclipse/thingweb.node-wot/issues/229. This makes it *easier* to identify where the logging comes from.

**Comment by egekorkan** [comment]
I can provide it in its complete form but basically it says serialized TD produced (like in the first expose) and then the TD. There is no mention of who logs it

**Comment by relu91** [comment]
> I would also support the proposal of #229.

I am working on it 👍

**Comment by danielpeintner** [comment]
> I am working on it 👍

Great, good to know. I also wanted to start working on it ;-)
I assume with the update we should also easily *see* where the TD is logged over and over again.

**Comment by relu91** [comment]
I found the source of this issue. Basically in the `HTTP-server` right before sending the answer to the observer is calling `thing.getThingDescription()`  ([link](https://github.com/eclipse/thingweb.node-wot/blob/b1f7774832f1988c6888dc367edd50910f17efc6/packages/binding-http/src/http-server.ts#L538) to the precise line) :
```ts
 try {
         let contentType = ProtocolHelpers.getPropertyContentType(thing.getThingDescription(), segments[3], "http");
         content = ContentSerdes.get().valueToContent(data, property.data, contentType);
        } catch(err) {
                console.warn("[binding-http]",`HttpServer on port ${this.getPort()} cannot process data for Event '${segments[3]}: ${err.message}'`);
                res.writeHead(500);
                res.end("Invalid Event Data");
                return;
       }
```
So far so good, the problem is that `thing.getThingDescription()` is calling `serializeTD(thing)`:
```ts
 public getThingDescription(): WoT.ThingDescription {
        return JSON.parse(TD.serializeTD(this));
    }
``` 
which at the ends outputs the famous line: `console.debug("[td-tools/td-parser]",`serializeTD() produced\n\`\`\`\n${td}\n\`\`\``);`. 

Now, one question and two possible solutions:
- *Question*: Why we are serializing the object only for parsing it again with `JSON.parse`?
- *Solution 1*: remove the line from `serializeTD`
- *Solution 2*: do not call `TD.serilize` and if sterilize is doing something fancy i.e. adding some properties/metadata store its output in a variable. 

Inputs are welcome 😃

**Comment by danielpeintner** [comment]
Question: With the recent logging level change to debug... is this still an issue?
Using default info logging the issue does not appear anymore, correct?

**Comment by relu91** [comment]
> Question: With the recent logging level change to debug... is this still an issue?
Using default info logging the issue does not appear anymore, correct?

Correct, it happens only with the debug level enabled. Anyway, it is quite odd to log the serialized TD every time `getThingDescription` is called.

**Comment by egekorkan** [comment]
I also think that the TD should not be outputted unless it changes like mentioned in solution 2. I would say that solution 1 is better since another use of serializeTD would actually need the TD outputted in a debug level.

**Comment by relu91** [comment]
> I would say that solution 1 is better since another use of serializeTD would actually need the TD outputted at a debug level.

yeah also to me, it feels like that logline should be written by the caller of `serializeTD` rather than `serializeTD` itself.

**Comment by danielpeintner** [comment]
Makes, sense and I am OK with
> Solution 1: remove the line from serializeTD

Please approve PR https://github.com/eclipse/thingweb.node-wot/pull/265 if ok

**Comment by danielpeintner** [comment]
see comment in PR https://github.com/eclipse/thingweb.node-wot/pull/265#pullrequestreview-441459338

> Perfect! However I'd still open an issue to keep track of why getThingDescription method calls serializeTD just to parse it again.

The only reason why `serializeTD()` is called is that it adds defaults and tries to clean-up the ThingDescription. Do I miss anything?

**Comment by danielpeintner** [comment]
Closed https://github.com/eclipse/thingweb.node-wot/issues/271 in favor of this issue to have a history of the issue.
